### PR TITLE
Moved core code form ikMail share extension.

### PR DESCRIPTION
**Pull request by adrien-coye**
### Misc

This contains: 
- Expanding ParallelTaskMapper to work with abstracted collections that match correct swift core protocols. (Sequence & Collection)
  - So we can use it with ArraySlice, without having to spown a new Array() each time
  - Also works for free with Dictionary now
  - Unit tested
- Closure kit, first small iteration
  - Define a base closure type
  - Can happend them, and execute them.
  - Unit tested

### CODE REVIEW

File: Sources/InfomaniakCore/ClosureKit/AsyncCurriedClosure.swift [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+
+/// Represents `any` (ie. all of them not the type) curried closure, of arbitrary type.
+///
+/// Supports concurrency and error
+@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public typealias AsyncCurriedClosure<Input, Output> = (Input) async throws -> Output
```
Comment (adrien-coye) [inline Sources/InfomaniakCore/ClosureKit/AsyncCurriedClosure.swift:25]: I think this has value, but we probaby need to discuss the  `~`operator and the call scheme of void closures as I try to keep it simple

File: Sources/InfomaniakCore/Extensions/Collection+Safe.swift [hunk 2]
```diff
@@ -24,3 +24,9 @@ public extension Collection {
         return indices.contains(index) ? self[index] : nil
     }
 }
+
```
Comment (adrien-coye) [inline Sources/InfomaniakCore/Extensions/Collection+Safe.swift:27]: Doc

File: Sources/InfomaniakCore/ClosureKit/AsyncCurriedClosure.swift (hunk 1 above)
Comment (adrien-coye) [reply Sources/InfomaniakCore/ClosureKit/AsyncCurriedClosure.swift:25]: Discussed, this should be removed.

File: Sources/InfomaniakCore/ClosureKit/CurriedClosure.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+
+/// Represents `any` (ie. all of them not the type) curried closure, of arbitrary type.
+///
+/// see `AsyncCurriedClosure` if you need support for structured concurrency or error feedback
+@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
```
Comment (adrien-coye) [inline Sources/InfomaniakCore/ClosureKit/CurriedClosure.swift:24]: Discussed: @available not necessary on this one, no async code.

**Review by adrien-coye — COMMENTED** [review]
Discussed with @PhilippeWeidmann

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**